addons/test_mail/tests/test_performance.py

```diff
@@ -854,7 +854,7 @@ def test_complex_message_post_template(self):
         template_id = self.env.ref('test_mail.mail_test_container_tpl').id
 
         # about 20 (19 ?) queries per additional customer group
-        with self.assertQueryCount(__system__=42, employee=43):
+        with self.assertQueryCount(__system__=42, employee=49):
```

Comment:
Adding a `active` on a Model can add extra queries because of `https://github.com/odoo/odoo/blob/master/odoo/fields.py#L4087` (will fetch `active` of companies when you use a XtoMany that target res.company, maybe user.company_ids in this case).

@beledouxdenis , you should change `_get_company_ids` (orm_cached method to have company_ids of the user) on `res.users` to be deterministic and immutable (a tuple maybe) ? Also, The `companies` (lazy_property) of `Environment` class should use it instead of `self.user.company_ids`. 
With you want to avoid any extra queries, you need to change every `self.user.company_ids` by `self.user._get_company_ids()`.

Reply:
Well this is an urgent request from Antony I received Today, which actually will target 16.0. I am doing it in master for the moment to avoid having to do an exception in the runbot for migration scripts.

So if I can avoid a large refactoring to be able to just add this active field, I would prefer :D. 

I am nevertheless having a look from where comes these extra requests.

Reply:
Thanks @ryv-odoo for the advice :)

Would be interesting to add an enterprise build as we also have query counter tests for payroll.

Reply:
Indeed at least stating in commit message where they come from, it helps trying to remove them afterwards when the OXP rush is finished, as it can be found in history of counter changes (most of them are undocumented hélàs).

Reply:
@beledouxdenis If I may add a comment, the runbot team prefers to add runbot exceptions instead of merging from master to the targeted stable branch, as this will break builds for other branches that are not up-to-date.

I can help you creating the exception if you need some assistance.

cc @Xavier-Do

Reply:
```diff
--- before.log  2022-10-07 16:16:49.541434600 +0200
+++ after.log   2022-10-07 16:17:10.985897600 +0200
@@ -12,7 +12,29 @@
  SELECT "mail_message".id FROM "mail_message" WHERE ((("mail_message"."res_id" = 1) AND ("mail_message"."model" = 'mail.test.container')) AND (("mail_message"."message_type" != 'user_notification') OR "mail_message"."message_type" IS NULL)) ORDER BY  "mail_message"."id" ASC  LIMIT 1
  SELECT "mail_alias".id FROM "mail_alias" LEFT JOIN "ir_model" AS "mail_alias__alias_model_id" ON ("mail_alias"."alias_model_id" = "mail_alias__alias_model_id"."id") WHERE ((("mail_alias"."alias_parent_model_id" in (SELECT "ir_model".id FROM "ir_model" WHERE ("ir_model"."model" = 'mail.test.container'))) AND ("mail_alias"."alias_parent_thread_id" in (1))) AND "mail_alias"."alias_name" IS NOT NULL) ORDER BY  COALESCE("mail_alias__alias_model_id"."is_mail_thread", false) DESC,"mail_alias__alias_model_id"."name"->>'en_US' ASC,"mail_alias"."alias_name"
  SELECT "mail_alias"."id" AS "id", "mail_alias"."alias_name" AS "alias_name", "mail_alias"."alias_model_id" AS "alias_model_id", "mail_alias"."alias_user_id" AS "alias_user_id", "mail_alias"."alias_defaults" AS "alias_defaults", "mail_alias"."alias_force_thread_id" AS "alias_force_thread_id", "mail_alias"."alias_parent_model_id" AS "alias_parent_model_id", "mail_alias"."alias_parent_thread_id" AS "alias_parent_thread_id", "mail_alias"."alias_contact" AS "alias_contact", "mail_alias"."alias_bounced_content"->>'en_US' AS "alias_bounced_content", "mail_alias"."create_uid" AS "create_uid", "mail_alias"."create_date" AS "create_date", "mail_alias"."write_uid" AS "write_uid", "mail_alias"."write_date" AS "write_date" FROM "mail_alias" WHERE "mail_alias".id IN (1)
- SELECT "res_company"."id" AS "id", "res_company"."name" AS "name", "res_company"."sequence" AS "sequence", "res_company"."parent_id" AS "parent_id", "res_company"."partner_id" AS "partner_id", "res_company"."report_header" AS "report_header", "res_company"."report_footer"->>'en_US' AS "report_footer", "res_company"."company_details" AS "company_details", "res_company"."currency_id" AS "currency_id", "res_company"."email" AS "email", "res_company"."phone" AS "phone", "res_company"."mobile" AS "mobile", "res_company"."paperformat_id" AS "paperformat_id", "res_company"."external_report_layout_id" AS "external_report_layout_id", "res_company"."base_onboarding_company_state" AS "base_onboarding_company_state", "res_company"."font" AS "font", "res_company"."primary_color" AS "primary_color", "res_company"."secondary_color" AS "secondary_color", "res_company"."layout_background" AS "layout_background", "res_company"."create_uid" AS "create_uid", "res_company"."create_date" AS "create_date", "res_company"."write_uid" AS "write_uid", "res_company"."write_date" AS "write_date", "res_company"."partner_gid" AS "partner_gid", "res_company"."iap_enrich_auto_done" AS "iap_enrich_auto_done", "res_company"."snailmail_color" AS "snailmail_color", "res_company"."snailmail_cover" AS "snailmail_cover", "res_company"."snailmail_duplex" AS "snailmail_duplex" FROM "res_company" WHERE ("res_company"."id" in (1)) AND "res_company".id IN (1)
+  SELECT r.id FROM ir_rule r JOIN ir_model m ON (r.model_id=m.id)
+                    WHERE m.model='res.company' AND r.active AND r.perm_read
+                    AND (r.id IN (SELECT rule_group_id FROM rule_group_rel rg
+                                  JOIN res_groups_users_rel gu ON (rg.group_id=gu.gid)
+                                  WHERE gu.uid=8)
+                         OR r.global)
+                    ORDER BY r.id
+
+  SELECT res_company_users_rel.user_id, res_company_users_rel.cid FROM res_company_users_rel, "res_company"
+                    WHERE 1=1 AND res_company_users_rel.user_id IN (8) AND res_company_users_rel.cid = res_company.id
+                     ORDER BY "res_company"."sequence" ,"res_company"."name"
+
+ SELECT "res_company"."id" AS "id", "res_company"."name" AS "name", "res_company"."active" AS "active", "res_company"."sequence" AS "sequence", "res_company"."parent_id" AS "parent_id", "res_company"."partner_id" AS "partner_id", "res_company"."report_header" AS "report_header", "res_company"."report_footer"->>'en_US' AS "report_footer", "res_company"."company_details" AS "company_details", "res_company"."currency_id" AS "currency_id", "res_company"."email" AS "email", "res_company"."phone" AS "phone", "res_company"."mobile" AS "mobile", "res_company"."paperformat_id" AS "paperformat_id", "res_company"."external_report_layout_id" AS "external_report_layout_id", "res_company"."base_onboarding_company_state" AS "base_onboarding_company_state", "res_company"."font" AS "font", "res_company"."primary_color" AS "primary_color", "res_company"."secondary_color" AS "secondary_color", "res_company"."layout_background" AS "layout_background", "res_company"."create_uid" AS "create_uid", "res_company"."create_date" AS "create_date", "res_company"."write_uid" AS "write_uid", "res_company"."write_date" AS "write_date", "res_company"."partner_gid" AS "partner_gid", "res_company"."iap_enrich_auto_done" AS "iap_enrich_auto_done", "res_company"."snailmail_color" AS "snailmail_color", "res_company"."snailmail_cover" AS "snailmail_cover", "res_company"."snailmail_duplex" AS "snailmail_duplex" FROM "res_company" WHERE "res_company".id IN (1)
+  SELECT res_groups_users_rel.uid, res_groups_users_rel.gid FROM res_groups_users_rel, "res_groups"
+                    WHERE 1=1 AND res_groups_users_rel.uid IN (8) AND res_groups_users_rel.gid = res_groups.id
+                     ORDER BY "res_groups"."name"->>'en_US'
+
+ SELECT "ir_rule"."id" AS "id", "ir_rule"."name" AS "name", "ir_rule"."active" AS "active", "ir_rule"."model_id" AS "model_id", "ir_rule"."domain_force" AS "domain_force", "ir_rule"."perm_read" AS "perm_read", "ir_rule"."perm_write" AS "perm_write", "ir_rule"."perm_create" AS "perm_create", "ir_rule"."perm_unlink" AS "perm_unlink", "ir_rule"."create_uid" AS "create_uid", "ir_rule"."create_date" AS "create_date", "ir_rule"."write_uid" AS "write_uid", "ir_rule"."write_date" AS "write_date", "ir_rule"."global" AS "global" FROM "ir_rule" WHERE "ir_rule".id IN (17)
+  SELECT rule_group_rel.rule_group_id, rule_group_rel.group_id FROM rule_group_rel, "res_groups"
+                    WHERE 1=1 AND rule_group_rel.rule_group_id IN (17) AND rule_group_rel.group_id = res_groups.id
+                     ORDER BY "res_groups"."name"->>'en_US'
+
+ SELECT "res_company"."id" AS "id", "res_company"."name" AS "name", "res_company"."active" AS "active", "res_company"."sequence" AS "sequence", "res_company"."parent_id" AS "parent_id", "res_company"."partner_id" AS "partner_id", "res_company"."report_header" AS "report_header", "res_company"."report_footer"->>'en_US' AS "report_footer", "res_company"."company_details" AS "company_details", "res_company"."currency_id" AS "currency_id", "res_company"."email" AS "email", "res_company"."phone" AS "phone", "res_company"."mobile" AS "mobile", "res_company"."paperformat_id" AS "paperformat_id", "res_company"."external_report_layout_id" AS "external_report_layout_id", "res_company"."base_onboarding_company_state" AS "base_onboarding_company_state", "res_company"."font" AS "font", "res_company"."primary_color" AS "primary_color", "res_company"."secondary_color" AS "secondary_color", "res_company"."layout_background" AS "layout_background", "res_company"."create_uid" AS "create_uid", "res_company"."create_date" AS "create_date", "res_company"."write_uid" AS "write_uid", "res_company"."write_date" AS "write_date", "res_company"."partner_gid" AS "partner_gid", "res_company"."iap_enrich_auto_done" AS "iap_enrich_auto_done", "res_company"."snailmail_color" AS "snailmail_color", "res_company"."snailmail_cover" AS "snailmail_cover", "res_company"."snailmail_duplex" AS "snailmail_duplex" FROM "res_company" WHERE ("res_company"."id" in (1)) AND "res_company".id IN (1)
  INSERT INTO "mail_message" ("author_guest_id", "author_id", "body", "create_date", "create_uid", "date", "email_add_signature", "email_from", "email_layout_xmlid", "mail_activity_type_id", "mail_server_id", "message_id", "message_type", "model", "parent_id", "record_name", "reply_to", "reply_to_force_new", "res_id", "subject", "subtype_id", "write_date", "write_uid") VALUES (NULL, 41, '<p>Adding stuff on Test Container</p>', '2022-10-07', 8, '2022-10-07', false, '"Ernest Employee" <e.e@example.com>', NULL, NULL, NULL, '<openerp-1-mail.test.container@>', 'comment', 'mail.test.container', 19, 'Test Container', '"YourCompany Test Container" <test-alias@test.com>', false, 1, 'Post on Test Container', 1, '2022-10-07', 8) RETURNING "id"
  INSERT INTO mail_message_res_partner_rel (mail_message_id, res_partner_id) VALUES (23, 43) ON CONFLICT DO NOTHING
  SELECT DISTINCT id, model, res_id, author_id, parent_id, message_type FROM "mail_message" WHERE id = ANY (ARRAY[23])
```

Reply:
It's looks to come from a caching issue on the `_compute_domain` of `ir.rule`. I have to leave now but will continue investigation later.

Reply:
Ok the issue was indeed coming from
https://github.com/odoo/odoo/blob/a9774ca9cbede99993b9931d0d53733165ae66e1/odoo/fields.py#L4087-L4089

Basically, when accessing the env lazy property `env.companies`, `self.user.company_ids` goes in the above code, and needs to fetch `active` (and take the opportunity to prefetch other fields) when applying `filtered` on the record set, to remove inactive records.

To overcome the issue, in `res.users._get_company_ids`, I replace the call `self.company_ids.ids`  to
```py
self.env['res.company'].search([('user_ids', 'in', self._uid)]).ids
```
This will filter the inactive records implicitly in the search, directly in SQL, rather than filtering the inactive records in python.
We spare the extra SQL queries that way.

That being said, the cache key of `_get_company_ids` should be changed to add `self.env.context.get('active_test')`, or within it, we should force it to `active_test=True`. I would tend to the second option, as it is highly unlikely we would like to include inactive companies through `env.companies` or through `res.users._get_company_ids`.

In addition, this means that any code doing `user.company_ids` will now do extra queries compared to before, so adding this `active` fields on `res.company` shouldn't be considered risk-free. Though, there are not so much `user.company_ids` (44 occurrences)